grudge/symbolic/operators.py

```diff
@@ -651,20 +658,41 @@ def matrix(self, afgrp, volgrp, dtype):
                     afgrp.nunit_dofs),
                 dtype=dtype)
 
-        from modepy.tools import UNIT_VERTICES
         import modepy as mp
-        for iface, fvi in enumerate(
-                volgrp.mesh_el_group.face_vertex_indices()):
-            face_vertices = UNIT_VERTICES[volgrp.dim][np.array(fvi)].T
-            matrix[:, iface, :] = mp.nodal_face_mass_matrix(
-                    volgrp.basis(), volgrp.unit_nodes, afgrp.unit_nodes,
-                    volgrp.order,
-                    face_vertices)
-
-        # np.set_printoptions(linewidth=200, precision=3)
-        # matrix[np.abs(matrix) < 1e-13] = 0
-        # print(matrix)
-        # 1/0
+        from meshmode.discretization import ElementGroupWithBasis
+
+        n = volgrp.order
+        m = afgrp.order
+        vol_basis = volgrp.basis_obj()
+        faces = mp.faces_for_shape(volgrp.shape)
+        face_quad_degree = 2*max(n, m)
+
+        for iface, face in enumerate(faces):
+            try:
+                face_quadrature = afgrp._quadrature_rule()
+            except AttributeError:
+                face_quadrature = mp.quadrature_for_space(
+                    mp.space_for_shape(face, face_quad_degree),
```

Comment:
Doesn't this lead to different quadrature orders?
* In the case where there is a `_quadrature_rule`, the order is `afgrp.order`
* In the case there isn't, the order is `2 * max(volgrp.order, afgrp.order)`

Reply:
In the case of this example in mirgecom: https://github.com/illinois-ceesd/mirgecom/blob/2a39aa880ff883774f9646e23746c682c151fd80/test/test_diffusion.py#L271
The over integration order on the `QuadratureSimplexElementGroup` I think is already taking into account what quadrature degree you want. (3*order). And so that's why I think we should use the underlying quadrature rule instead of creating one.

My understanding of these quadrature element groups is that the order is an explicit statement on the order of quadrature you want to use. Is this a correct statement?

Now the `except` clause is more or less handling the "non over integrated" case, with possibly different quadrature rules in the volume and face. And the calculation for the order ensures that we integrate the product of polynomials restricted to the face exactly, which is more or less what the "old" face mass routine in modepy does: https://github.com/inducer/modepy/blob/7ebb43bd7612adf55991c5535811d0eaf8f4368e/modepy/matrices.py#L358

So this is just a version of that which considers the possibility that face and volume quadrature may have different orders.

After thinking about this as I'm writing, I'm wondering if a more readable way is to do something along the lines of: `if isinstance(afgrp, QuadratureSimplexElementGroup):...` rather than try-except. I think that's technically equivalent to the try-except code here, but maybe it's more explicit?

Reply:
> My understanding of these quadrature element groups is that the order is an explicit statement on the order of quadrature you want to use. Is this a correct statement?

No. `order` tends to serve as a loose identifier of different rules within a family of quadrature rules. It certainly correlates with exactness degree, and sometimes that's exactly what you get, but in general you have to look at the `exact_to` attribute to get a promise of exactness.

My sense is that we should just recast the `NodalElementGroupBase` to make a modepy quadrature rule available (and, consequently, deprecate the `weights` attribute).

> But it feels wasteful to keep instantiating new mp.Quadrature objects when there is one already available.

Instantiating objects is cheap. Recomputing weights is not. (But see above.)

Reply:
> No. `order` tends to serve as a loose identifier of different rules within a family of quadrature rules. It certainly correlates with exactness degree, and sometimes that's exactly what you get, but in general you have to look at the `exact_to` attribute to get a promise of exactness.


OK. Thanks for the clarification.


> My sense is that we should just recast the `NodalElementGroupBase` to make a modepy quadrature rule available (and, consequently, deprecate the `weights` attribute).


OK. So basically make `_quadrature_rule` public?


> Instantiating objects is cheap. Recomputing weights is not. (But see above.)


Good to know.

Reply:
Uhg. Not all `modepy.Quadrature` objects have an `exact_to` method...

Reply:
> OK. So basically make `_quadrature_rule` public?

Well, yes, but also move it up to `NodalElementGroupBase`.

Reply:
> Edit: Oh. I'm getting smacked by the zero-dimensional quadrature rule again.

@thomasgibson asked in chat what the accuracy should be. I wasn't sure. `np.inf`? `None`?

Also, it might be worthwhile to centralize/standardize that 0D quadrature rule in modepy, so that we don't create multiple inconsistent versions of it.

Reply:
> `np.inf`? `None`?

I kind of like `np.inf`. It makes sense, comparisons with some finite `exact_to` are valid and there's already the special case of `AttributeError` to say we don't know.

Reply:
What about a general quadrature rule constructed from a set of weights/nodes? To me, this seems almost impossibly difficult to determine. Which is why I get why there is no generic `exact_to` method for `modepy.Quadrature`. Perhaps what should happen is that `modepy.Quadrature` should not be allowed to be instantiated on its own (i.e. make it an abstract class); rather have only concrete subclass of it, where we actually know about the quad rule and where it comes from.

Reply:
Yes, agreed on all of that. 

I was thinking more about adding a separate `PointQuadrature` for 0d things, which can have an `np.inf` accuracy.

Reply:
> To me, this seems almost impossibly difficult to determine.

*What* seems almost impossibly difficult to determine?

Reply:
That sounds like a good idea to me @alexfikl!

@inducer I guess I should have been more clear; what I am talking about is if someone instantiates an instance of `modepy.Quadrature` by passing in some set of weights and nodes. Given those two bits of info, _what is the order of accuracy_? What I then suggested is:

> ... `modepy.Quadrature` should not be allowed to be instantiated on its own (i.e. make it an abstract class); rather have only concrete subclass of it

That's one proposed solution, which is in line with what @alexfikl is suggesting about having a concrete type for 0-D quadrature. That way, we don't have to play guessing games to figure out what `exact_to` should do.

Granted, this is not a huge problem for most rules. But since we want to check the order of accuracy of the quadrature rule when assembling the face mass matrix, we should make sure we can always get that information in a clear, simple way.

That's it.

Reply:
To give credit, I was just echoing what @inducer said before

> Also, it might be worthwhile to centralize/standardize that 0D quadrature rule in modepy, so that we don't create multiple inconsistent versions of it.

I'm also ok with `mp.Quadrature` not being an abstract class and just acting as a bag of nodes and weights that the user promises is a quadrature rule. Similar to the `PolynomialGivenNodesElementGroup` thing in meshmode.

Reply:
Thanks for explaining!

> That way, we don't have to play guessing games to figure out what `exact_to` should do.

I agree that we shouldn't try to guess `exact_to` from provided nodes and weights, it should be supplied, by someone.

* For the 0D quadrature rule, making it its own class will neatly solve this issue. The class can define an `exact_to` attribute, and we're done.
* For other quadratures, I think it's important to keep apart the notion of exactness, and whether the rule exists as a separate class. It's very possible to make a quadrature subclass that receives `exact_to` data via its constructor argument.
* Unrelated: https://github.com/inducer/modepy/issues/31